bpo-44938: Expose PyErr_ChainExceptions in the stable API (#27799)
python/cpython: Doc/c-api/exceptions.rst

   .. versionadded:: 3.5

.. c:function:: void PyErr_ChainExceptions(PyObject *exc, PyObject *val, PyObject *tb)

>    Set the provided exception as the cause for the current exception. If there is no exception

Reviewer: Does this really do cause? Where?

Reviewer: There are only two exception here: the "current exception" C and the "provided exception" P.   This function sets C to be the context of P. The it makes P the new "current exception". Cause is not set.

Author: > Does this really do cause? Where?

That is refering to the `exception.__cause__` argument and related machinery that you can retrieve with `https://docs.python.org/3/c-api/exceptions.html#c.PyException_GetCause`

Reviewer: I don't think this function takes a __cause__ argument. Check the code.

Author: > I don't think this function takes a **cause** argument. Check the code.

I am confused, unfortunately. The text says:

> Set the provided exception as the cause for the current exception. 

~So the provided exception (first argument) is set as the `__cause__` for the current exception~

Author: Ah, my bad, I am mixing things. Will clarify this sentence

Author: Just to clarify the confusion: I was thinking of `__context__`, not `__cause__`. The latter is for `raise ... from ...`

   currently set (:c:func:`PyErr_Occurred` returns ``0``) this function behaves as like :c:func:`PyErr_Restore`, but if an exception is already set, set the context

>    associated with it.

Reviewer: I'm not sure what you mean by "if an exception is already set, set the context associated with it".

Author: > "if an exception is already set, set the context associated with it".

The context is part of the exception chaining mechanism (is what is stored in `f.__context__`). Check also https://docs.python.org/3/c-api/exceptions.html#c.PyException_GetContext

Reviewer: I meant that the wording of this sentence doesn't seem clear to me.  In particular, what is "it" - is it where you set or is it what the context is associated with?

Author: > I meant that the wording of this sentence doesn't seem clear to me.

Ah, gotcha. Let me rewrite this a bit to see if I can get it a bit clearer

Author: Check the latest commit. The sentence may read a bit complex in English so any improvement or suggestion on the structure is welcomed :)

>    The exception type should be a class. Do not pass an invalid exception type or value. This call

Reviewer: What happens if I do pass an invalid exception type?

Author: The same as if you pass an invalid exception type to `PyErr_Restore` that line is adapted from there: `PyExc_SystemError` will be set as the exception. 

But I don't feel we want to make that part of the public contract...yet.

   takes away a reference to each object: you must own a reference to each object before the call

>    and after the call you no longer own these references.

Reviewer: You could just replace the whole sentence by "This function steals a reference to each of its arguments" like in other places.

   .. warning::

      The caller is responsible for ensuring that this call won't create

>       any cycles in the exception context chain.

Reviewer: I think the doc should say what happens if you violate the rule.  I think in reality it will just create the cycle and if it doesn't bother you then good for you.

Author: > I think the doc should say what happens if you violate the rule.

Is undefined behavior, technically as the contract of this function withing CPython is that cycles should not be created. Some parts of the exception machinery can break (or at least they used to) in presence of cycles.

Reviewer: But when the exception machinery breaks in the presence of a cycle we consider that to be a bug in python.

Reviewer 2: How is the caller supposed to ensure this? Is there any API for checking it?

Author: There is no API to check this, but to create a cycle with exceptions it needs to be s deliberate action. There is not an easy way to accidentally crate an exception cycle

Reviewer 3: > But when the exception machinery breaks in the presence of a cycle we consider that to be a bug in python.

I think the movement is in the direction of trying to prevent / avoid the creation of cycles in the first place (as was mentioned in the discussion for this [recent ticket](https://bugs.python.org/issue25782), for example). The warning being added here is in line with that.

Reviewer 4: As per my comment below, the easiest way to violate this requirement is to call the API without setting a new exception first (or clearing the existing one). Buggy code (using my API naming suggestion from above):

```
PyErr_Fetch(&type, &value, &traceback);
// Exception state is not cleared, no new exception is set
PyErr_RestoreOrChain(type, value, traceback); // Sets exception as its own context
```

Because the exception might not have been instantiated yet when `PyErr_Fetch()` is called, there's no easy way for the API to reliably detect this error, so the onus is on the caller to manage the exception state correctly.

   .. versionadded:: 3.5

.. c:function:: void PyErr_ChainExceptions(PyObject *exc, PyObject *val, PyObject *tb)

   Set the provided exception as the cause for the current exception. If there is no exception
   currently set (:c:func:`PyErr_Occurred` returns ``0``) this function behaves
   as like :c:func:`PyErr_Restore`, but if an exception is already set, set the
   context associated with it.

   The exception type should be a class. Do not pass an invalid exception type or value. This
   function steals a reference to each of its arguments in the same fashion as :c:func:`PyErr_Restore`
   does.

   .. warning::

      The caller is responsible for ensuring that this call won't create
      any cycles in the exception context chain.

   See :ref:`tut-exception-chaining` for more information regarding the semantics of exception
   chaining.

   Example usage:

   .. code-block:: c

      PyObject *type, *value, *traceback;

      /* Fetch the original exception */
      PyErr_Fetch(&type, &value, &traceback);

      /* ... code that might produce other errors ... */

>       /* Set the original exception as the cause for the new exception */

Reviewer: ```suggestion
      /* Set the original exception as the context for the new exception, and set the new exception as the current exception */
```

   .. versionadded:: 3.5

.. c:function:: void PyErr_ChainExceptions(PyObject *exc, PyObject *val, PyObject *tb)

>    Set the provided exception as the cause for the current exception. If there

Reviewer: ```suggestion
   Set the provided exception as the context for the current exception. If there
```

   is no exception currently set (:c:func:`PyErr_Occurred` returns ``0``) this

>    function behaves as like :c:func:`PyErr_Restore`, but if an exception is

Reviewer 2: ```suggestion
   function behaves like :c:func:`PyErr_Restore`, but if an exception is
```

   already set, set the provided exception as the context (accessible from
   Python through the :attr:`__context__` attribute of the exception) of the

>    one that is currently set.

Reviewer: Isn't it the other way around - the existing exception becomes the context of the provided exception?

Author: The code calls `PyException_SetContext(val2, val);` where `val` is the provided one and `val2` is the one that is currently set. `PyException_SetContext` sets the second argument as the context for the first.

So unless I am missing something (is late here :sweat_smile: ) this means that "The provided exception (`val`) becomes the context of the existing exception (`val2`)"

Reviewer: Right, I was thinking we are raising the provided exception in the context of the existing one, but that's what SetObject is for.   Can we think of a more descriptive name for this function? I confused the direction, and you confused cause/context because "chain" is ambiguous here.

How about PyErr_RestoreContext?   IIUC this describes what you use this function for - putting back a fetched exception, as the context of any exception that happened between the fetch and now.  (Should something like the previous sentence be in the doc?)

Reviewer: Or possibly PyErr_RestoreAsContext?

Author: In general authors look for "chain" in the docs because that's the action they want to do, so I would really recommend to have "chain" in the name.

Reviewer: That may be the case, but I think we should promote clear language, and "chain" in this context is vague. The doc can contain "chain" in the prose for the benefit of people who search for it (for now). But should it be part of the function name for all eternity?

Author: Is unclear to me. I agree with the ambiguity but what users if the function want is a higher level approach to the python behaviour: an exception is raised when another exception was there and I want to do in C the same thing as Python does.

I'm not a native speaker do I trust your judgement regarding how ambiguous this sounds and that we need a better name so let's search for one.

Maybe we should make a section in the docs about "exception chaining" and offer two APIs: one for cause and another for context

Reviewer 3: @iritkatriel Maybe you can suggest a more specific name that still retains "chain" as part of it?

Reviewer: The context is an implicitly chained exception (the cause is explicitly chained). There is no need to have both chained and context in the name. If we have only one of them I think it should be context and not chain.

Reviewer 4: Given the equivalence  to `PyErr_Restore` when no exception is currently set, my public naming suggestion would be `PyErr_RestoreOrChain`, with the shorthand documentation being "Restores the given exception state, or chains it as the exception context if an exception is already set".

The key point to emphasise is that the function takes the same arguments as `PyErr_Restore` (retrieved from a prior `PyErr_Fetch` call), and generally does the same thing, but handles currently set exceptions differently: instead of clearing them, it leaves them in place, and "restores" the given exception as the exception context, rather than as the active exception.

   The exception type should be a class. Do not pass an invalid exception type

>    or value. This function steals a reference to each of its arguments in the

Reviewer 2: Could it *not* steal its arguments?
Functions that don't follow the usual calling convention are harder to reason about; it would be good to treat them as historical exceptions.

Author: It steals the arguments because of what it calls. It had the same calling convention as `Pyerr_Restore`.

Author: Also, stealing arguments is not an uncommon calling convention.

Reviewer 4: ~Stealing references is only common because much of the public API was born as internal APIs, and we often try to avoid redundant refcount adjustments in internal APIs.~

~It took us a long time to realise that stealing references in public APIs makes things confusing for API clients, and can easily lead to double decref bugs (which are more potentially destructive than the object leaks that can arise from missing decref bugs, since deleting objects early can lead to access violations). So I think the new public API should be a wrapper around the internal API that avoids stealing references.~

I changed my mind on this point, as `PyErr_Fetch` and `PyErr_Restore` are a matched pair. If we go with my `PyErr_RestoreOrChain` naming suggestion, then the refcounting semantics of the new API should match those of `PyErr_Restore` (which is to take back the references that `PyErr_Fetch` created)

   same fashion as :c:func:`PyErr_Restore` does.

   .. warning::

      The caller is responsible for ensuring that this call won't create
      any cycles in the exception context chain.

   See :ref:`tut-exception-chaining` for more information regarding the
   semantics of exception chaining.

   Example usage:

   .. code-block:: c

      PyObject *type, *value, *traceback;

      /* Fetch the original exception */
      PyErr_Fetch(&type, &value, &traceback);

      /* ... code that might produce other errors ... */

      /* Set the original exception as the context for the new exception, and set the new exception as the current exception */

>       PyErr_ChainExceptions(type, value, traceback);

Reviewer 4: This example is misleading, as if the code *doesn't* set a new exception, the fetched exception will end up being chained with itself instead of being restored. To safely invoke the new API unconditionally, the code needs to look like this:

```
    /* Fetch and clear the original exception */
    PyErr_Fetch(&type, &value, &traceback);
    PyErr_Clear();

    /* ... code that might produce other errors ... */

    /* Either restore the original exception, or set it as the context for any new exception */
    PyErr_RestoreOrChain(type, value, traceback);
```

Skipping the `PyErr_Clear()` call is only OK if the code *definitely* sets a new active exception.
